pkg/util/modules/modules.go

```diff
@@ -0,0 +1,157 @@
+package modules
+
+import (
+	"fmt"
+
+	"github.com/cortexproject/cortex/pkg/util/services"
+	"github.com/pkg/errors"
+)
+
+type service func() (services.Service, error)
+
+// module is the basic building block of the application
+type module struct {
+	// dependencies of this module
+	deps []string
+
+	// initFn for this module (can return nil)
+	initFn service
+}
+
+// Manager is a component that initialises modules of the application
+// in the right order of dependencies.
+type Manager struct {
+	modules map[string]module
+}
+
+// NewManager creates a new Manager
+func NewManager() *Manager {
+	return &Manager{
+		modules: make(map[string]module),
+	}
+}
+
+// RegisterModule registers a new module with ModuleManager
+func (m *Manager) RegisterModule(name string, initFn service) {
+	m.modules[name] = module{
+		initFn: initFn,
+	}
+	return
+}
+
+// AddDependency adds a dependency from name(source) to dependsOn(targets)
+func (m *Manager) AddDependency(name string, dependsOn ...string) error {
+	if mod, ok := m.modules[name]; ok {
+		for _, dep := range dependsOn {
+			if _, ok := m.modules[dep]; ok {
+				mod.deps = append(mod.deps, dep)
+			} else {
+				return fmt.Errorf("no such module: %s", dep)
+			}
+		}
+	} else {
+		return fmt.Errorf("no such module: %s", name)
+	}
+	return nil
+}
+
+// InitModuleServices starts the target module
+func (m *Manager) InitModuleServices(target string) (map[string]services.Service, error) {
+	servicesMap := map[string]services.Service{}
+
+	// initialize all of our dependencies first
+	deps := m.orderedDeps(target)
+	deps = append(deps, target) // lastly, initialize the requested module
+
+	for ix, n := range deps {
+		mod := m.modules[n]
+
+		var serv services.Service
+
+		if mod.initFn != nil {
+			s, err := mod.initFn()
+			if err != nil {
+				return nil, errors.Wrap(err, fmt.Sprintf("error initialising module: %s", n))
+			}
+
+			invDeps := m.findInverseDependencies(n, deps[ix+1:])
+			if s == nil {
+				if invDeps != nil {
+					return nil, fmt.Errorf("module %s returned nil service but has other modules dependent on it", n)
```

Comment:
Previously it was easier to find "inverse dependencies" in the source code, but in this dynamic version, it may be tricky. I think we should include them in the error message.

Reply:
I'm not sure I fully understand what you mean here.

Reply:
I mean to include names of "other modules"  in the error message. Basically.

```suggestion
					return nil, fmt.Errorf("module %s returned nil service but has other modules dependent on it: %v", n, invDeps)
```

Also it's better to check slice for length than `nil` (`if len(invDeps) > 0 ...`).